File: src/main/java/no/nav/veilarbaktivitet/aktivitetskort/service/AktivitetskortService.java
package no.nav.veilarbaktivitet.aktivitetskort.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.veilarbaktivitet.aktivitet.AktivitetDAO;
import no.nav.veilarbaktivitet.aktivitet.AktivitetService;
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData;
import no.nav.veilarbaktivitet.aktivitet.domain.Ident;
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetsMessageDAO;
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortCompareUtil;
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortMapper;
import no.nav.veilarbaktivitet.aktivitetskort.dto.bestilling.AktivitetskortBestilling;
import no.nav.veilarbaktivitet.aktivitetskort.dto.bestilling.ArenaAktivitetskortBestilling;
import no.nav.veilarbaktivitet.aktivitetskort.dto.bestilling.EksternAktivitetskortBestilling;
import no.nav.veilarbaktivitet.aktivitetskort.feil.AktivitetsKortFunksjonellException;
import no.nav.veilarbaktivitet.aktivitetskort.feil.ManglerOppfolgingsperiodeFeil;
import no.nav.veilarbaktivitet.aktivitetskort.feil.UlovligEndringFeil;
import no.nav.veilarbaktivitet.person.Person;
import no.nav.veilarbaktivitet.util.DateUtils;
import org.springframework.stereotype.Service;

import java.util.Objects;
import java.util.UUID;
import java.util.stream.Stream;


@Service
@RequiredArgsConstructor
@Slf4j
public class AktivitetskortService {

    private final AktivitetService aktivitetService;
    private final AktivitetDAO aktivitetDAO;
    private final AktivitetsMessageDAO aktivitetsMessageDAO;
    private final ArenaAktivitetskortService arenaAktivitetskortService;

    private final OppfolgingsperiodeService oppfolgingsperiodeService;


    public UpsertActionResult upsertAktivitetskort(AktivitetskortBestilling bestilling) throws AktivitetsKortFunksjonellException {
        var aktivitetskort = bestilling.getAktivitetskort();
        var gammelAktivitetVersjon = aktivitetDAO.hentAktivitetByFunksjonellId(aktivitetskort.getId());

        if (gammelAktivitetVersjon.isPresent()) {
            // Arenaaktiviteter er blitt "ekstern"-aktivitet etter de har blitt opprettet
            var oppdatertAktivitet = oppdaterAktivitet(gammelAktivitetVersjon.get(), bestilling.toAktivitet());
            log.info("Oppdaterte ekstern aktivitetskort {}", oppdatertAktivitet);
            return UpsertActionResult.OPPDATER;
        } else {
            var opprettetAktivitet = opprettAktivitet(bestilling);
            log.info("Opprettet ekstern aktivitetskort {}", opprettetAktivitet);
            return UpsertActionResult.OPPRETT;
        }
    }

    private AktivitetData opprettAktivitet(AktivitetskortBestilling bestilling) throws ManglerOppfolgingsperiodeFeil {
        if (bestilling instanceof ArenaAktivitetskortBestilling arenaAktivitetskortBestilling) {
            return arenaAktivitetskortService.opprettAktivitet(arenaAktivitetskortBestilling);
        } else if (bestilling instanceof EksternAktivitetskortBestilling eksternAktivitetskortBestilling) {
            return opprettEksternAktivitet(eksternAktivitetskortBestilling);
        } else {
            throw new IllegalStateException("Unexpected value: " + bestilling);
        }
    }

    private AktivitetData opprettEksternAktivitet(EksternAktivitetskortBestilling bestilling) throws ManglerOppfolgingsperiodeFeil {
        var endretAv = bestilling.getAktivitetskort().getEndretAv();
        var opprettet = bestilling.getAktivitetskort().getEndretTidspunkt().toLocalDateTime();
        var oppfolgingsperiode = oppfolgingsperiodeService.finnOppfolgingsperiode(bestilling.getAktorId(), opprettet);
        if (oppfolgingsperiode == null) throw new ManglerOppfolgingsperiodeFeil();

        var aktivitetData = AktivitetskortMapper
                .mapTilAktivitetData(bestilling, bestilling.getAktivitetskort().getEndretTidspunkt());

        if (oppfolgingsperiode.sluttDato() != null) {
            aktivitetData.getEksternAktivitetData().setOpprettetSomHistorisk(true);
            aktivitetData.getEksternAktivitetData().setOppfolgingsperiodeSlutt(oppfolgingsperiode.sluttDato().toLocalDateTime());
        }
        return aktivitetService.opprettAktivitet(Person.aktorId(aktivitetData.getAktorId()), aktivitetData, endretAv, opprettet, oppfolgingsperiode.uuid());
    }

    private AktivitetData oppdaterDetaljer(AktivitetData aktivitet, AktivitetData nyAktivitet) {
        if (AktivitetskortCompareUtil.erFaktiskOppdatert(nyAktivitet, aktivitet)) {
            return aktivitetService.oppdaterAktivitet(aktivitet, nyAktivitet, Person.navIdent(nyAktivitet.getEndretAv()), DateUtils.dateToLocalDateTime(nyAktivitet.getEndretDato()));
        }
        return aktivitet;
    }
    public AktivitetData oppdaterStatus(AktivitetData aktivitet, AktivitetData nyAktivitet) {
        if (aktivitet.getStatus() != nyAktivitet.getStatus()) {
            return aktivitetService.oppdaterStatus(
                aktivitet,
                nyAktivitet, // TODO: Populer avbrutt-tekstfelt
                new Ident(nyAktivitet.getEndretAv(), nyAktivitet.getEndretAvType()),
                DateUtils.dateToLocalDateTime(nyAktivitet.getEndretDato())
            );
        } else {
            return aktivitet;
        }
    }

    private AktivitetData oppdaterAktivitet(AktivitetData gammelAktivitet, AktivitetData nyAktivitet) throws UlovligEndringFeil {
        if (!Objects.equals(gammelAktivitet.getAktorId(), nyAktivitet.getAktorId())) throw new UlovligEndringFeil("Kan ikke endre bruker på samme aktivitetskort");
        if (gammelAktivitet.getHistoriskDato() != null) throw new UlovligEndringFeil("Kan ikke endre aktiviteter som er historiske (avsluttet oppfølgingsperiode)");
        //TODO vurder avtalt-logikken https://trello.com/c/dFyre4EK
        if (gammelAktivitet.isAvtalt() && !nyAktivitet.isAvtalt()) throw new UlovligEndringFeil("Kan ikke oppdatere fra avtalt til ikke-avtalt");

        return Stream.of(gammelAktivitet)
                .map( aktivitet -> settAvtaltHvisAvtalt( aktivitet, nyAktivitet))
                .map( aktivitet -> oppdaterDetaljer(aktivitet, nyAktivitet))
                .map( aktivitet -> oppdaterStatus(aktivitet, nyAktivitet))
                .findFirst().orElse(null);
    }

    private AktivitetData settAvtaltHvisAvtalt(AktivitetData originalAktivitet, AktivitetData nyAktivitet) {
        if (nyAktivitet.isAvtalt() && !originalAktivitet.isAvtalt()) {
            return aktivitetService.settAvtalt(
                    originalAktivitet,
                    new Ident(nyAktivitet.getEndretAv(), nyAktivitet.getEndretAvType()),
                    DateUtils.dateToLocalDateTime(nyAktivitet.getEndretDato()));
        } else {
            return originalAktivitet;
        }
    }

    public boolean harSettMelding(UUID messageId) {
        return aktivitetsMessageDAO.exist(messageId);
    }

    public void lagreMeldingsId(UUID messageId, UUID funksjonellId) {
        aktivitetsMessageDAO.insert(messageId, funksjonellId);
    }

    public void oppdaterMeldingResultat(UUID messageId, UpsertActionResult upsertActionResult, String reason) {
        aktivitetsMessageDAO.updateActionResult(messageId, upsertActionResult, reason);
    }

}

File: src/main/kotlin/no/nav/veilarbaktivitet/aktivitetskort/service/AktivitetskortService.kt
package no.nav.veilarbaktivitet.aktivitetskort.service

import no.nav.veilarbaktivitet.aktivitet.AktivitetDAO
import no.nav.veilarbaktivitet.aktivitet.AktivitetService
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetData
import no.nav.veilarbaktivitet.aktivitet.domain.Ident
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetsMessageDAO
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortCompareUtil
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortConsumer.OffsetAndPartition
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortMapper.toAktivitetsDataInsert
import no.nav.veilarbaktivitet.aktivitetskort.AktivitetskortMapper.toAktivitetsDataUpdate
import no.nav.veilarbaktivitet.aktivitetskort.bestilling.AktivitetskortBestilling
import no.nav.veilarbaktivitet.aktivitetskort.bestilling.ArenaAktivitetskortBestilling
import no.nav.veilarbaktivitet.aktivitetskort.bestilling.EksternAktivitetskortBestilling
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.MessageSource
import no.nav.veilarbaktivitet.aktivitetskort.feil.AktivitetsKortFunksjonellException
import no.nav.veilarbaktivitet.aktivitetskort.feil.ManglerOppfolgingsperiodeFeil
import no.nav.veilarbaktivitet.aktivitetskort.feil.UlovligEndringFeil
import no.nav.veilarbaktivitet.oppfolging.periode.IngenGjeldendePeriodeException
import no.nav.veilarbaktivitet.oppfolging.periode.SistePeriodeService
import no.nav.veilarbaktivitet.util.DateUtils
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*

@Service
class AktivitetskortService(
    private val aktivitetService: AktivitetService,
    private val aktivitetDAO: AktivitetDAO,
    private val aktivitetsMessageDAO: AktivitetsMessageDAO,
    private val arenaAktivitetskortService: ArenaAktivitetskortService,
    private val sistePeriodeService: SistePeriodeService
) {
    private val log = LoggerFactory.getLogger(javaClass)
    @Throws(AktivitetsKortFunksjonellException::class)
    fun upsertAktivitetskort(bestilling: AktivitetskortBestilling): UpsertActionResult {
        val (id) = bestilling.aktivitetskort
        // Splittede aktiviteter bestilles på ny funksjonell-id
        val gammelAktivitetMaybe = aktivitetDAO.hentAktivitetByFunksjonellId(id)
        return when {
            gammelAktivitetMaybe.isPresent -> {
                val gammelAktivitet = gammelAktivitetMaybe.get()
                val nyAktivitet = bestilling.toAktivitetsDataUpdate()
                // Arenaaktiviteter blir vanlig "ekstern"-aktivitet etter de har blitt opprettet
                val oppdatertAktivitet = when  {
                    bestilling is ArenaAktivitetskortBestilling && gammelAktivitet.erTattOverAvAnnetTeam() -> {
                        // Vi gjør arena-migrering men beholder andre team sine data, ikke data fra ACL, dette er pga en race-condition
                        // hvor andre team behandler endringer fra arena raskere enn oss
                        arenaAktivitetskortService.dobbelsjekkMigrering(bestilling, gammelAktivitet)
                        return UpsertActionResult.IGNORER
                    }
                    bestilling is ArenaAktivitetskortBestilling -> oppdaterAktivitet(gammelAktivitet, nyAktivitet, true)
                    else -> oppdaterAktivitet(gammelAktivitet, nyAktivitet, false)
                }
                log.info("Oppdaterte ekstern aktivitetskort {}", oppdatertAktivitet)
                UpsertActionResult.OPPDATER
            }
            else -> {
                val opprettetAktivitet = opprettAktivitet(bestilling)
                log.info("Opprettet ekstern aktivitetskort {}", opprettetAktivitet)
                UpsertActionResult.OPPRETT
            }
        }
    }

    @Throws(ManglerOppfolgingsperiodeFeil::class)
    private fun opprettAktivitet(bestilling: AktivitetskortBestilling): AktivitetData? {
        return when (bestilling) {
            is ArenaAktivitetskortBestilling -> arenaAktivitetskortService.opprettAktivitet(bestilling)
            is EksternAktivitetskortBestilling -> opprettEksternAktivitet(bestilling)
            else -> throw IllegalStateException("Unexpected value: $bestilling")
        }
    }

    @Throws(ManglerOppfolgingsperiodeFeil::class)
    private fun opprettEksternAktivitet(bestilling: EksternAktivitetskortBestilling): AktivitetData {
        val oppfolgingsperiode = try {
            sistePeriodeService.hentGjeldendeOppfolgingsperiodeMedFallback(bestilling.aktorId)
        } catch (e: IngenGjeldendePeriodeException) {
            throw ManglerOppfolgingsperiodeFeil()
        }
        val aktivitetData: AktivitetData = bestilling.toAktivitetsDataInsert()
        return aktivitetService.opprettAktivitet(
            aktivitetData.withOppfolgingsperiodeId(oppfolgingsperiode),
        )
    }

    private fun oppdaterDetaljer(aktivitet: AktivitetData, nyAktivitet: AktivitetData): AktivitetData {
        return if (AktivitetskortCompareUtil.erFaktiskOppdatert(nyAktivitet, aktivitet)) {
            aktivitetService.oppdaterAktivitet(
                aktivitet,
                nyAktivitet
            )
        } else aktivitet
    }

    fun oppdaterStatus(aktivitet: AktivitetData, nyAktivitet: AktivitetData): AktivitetData {
        return if (aktivitet.status != nyAktivitet.status) {
            aktivitetService.oppdaterStatus(
                aktivitet,
                nyAktivitet
            )
        } else {
            aktivitet
        }
    }

    @Throws(UlovligEndringFeil::class)
    private fun oppdaterAktivitet(gammelAktivitet: AktivitetData, nyAktivitet: AktivitetData, arenaAclOppdatering: Boolean): AktivitetData {
        if (gammelAktivitet.aktorId != nyAktivitet.aktorId) throw UlovligEndringFeil("Kan ikke endre bruker på samme aktivitetskort")
        // Arena-ACL kan foreløpig oppdatere historiske kort
        if (gammelAktivitet.historiskDato != null && !arenaAclOppdatering) throw UlovligEndringFeil("Kan ikke endre aktiviteter som er historiske (avsluttet oppfølgingsperiode)")
        if (gammelAktivitet.isAvtalt && !nyAktivitet.isAvtalt) throw UlovligEndringFeil("Kan ikke oppdatere fra avtalt til ikke-avtalt")
        return gammelAktivitet
            .let { aktivitet: AktivitetData -> settAvtaltHvisAvtalt(aktivitet, nyAktivitet) }
            .let { aktivitet: AktivitetData -> oppdaterDetaljer(aktivitet, nyAktivitet) }
            .let { aktivitet: AktivitetData -> oppdaterStatus(aktivitet, nyAktivitet) }
    }

    private fun settAvtaltHvisAvtalt(originalAktivitet: AktivitetData, nyAktivitet: AktivitetData): AktivitetData {
        return if (nyAktivitet.isAvtalt && !originalAktivitet.isAvtalt) {
            aktivitetService.settAvtalt(
                originalAktivitet,
                Ident(nyAktivitet.endretAv, nyAktivitet.endretAvType),
                DateUtils.dateToLocalDateTime(nyAktivitet.endretDato)
            )
        } else {
            originalAktivitet
        }
    }

    fun harSettMelding(messageId: UUID): Boolean {
        return aktivitetsMessageDAO.exist(messageId)
    }

    fun lagreMeldingsId(messageId: UUID, funksjonellId: UUID, offsetAndPartition: OffsetAndPartition) {
        aktivitetsMessageDAO.insert(messageId, funksjonellId, offsetAndPartition)
    }

    fun oppdaterMeldingResultat(messageId: UUID, upsertActionResult: UpsertActionResult, reason: String?) {
        aktivitetsMessageDAO.updateActionResult(messageId, upsertActionResult, reason)
    }

    fun hentAktivitetskortByFunksjonellId(funksjonellId: UUID): Optional<AktivitetData> {
        return aktivitetDAO.hentAktivitetByFunksjonellId(funksjonellId)
    }
}

fun AktivitetData.erTattOverAvAnnetTeam() = this.eksternAktivitetData.source != MessageSource.ARENA_TILTAK_AKTIVITET_ACL.name